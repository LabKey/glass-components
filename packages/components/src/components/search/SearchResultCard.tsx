/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react'
import { SCHEMAS } from '../base/models/schemas';
import { SVGIcon } from '../base/SVGIcon';

interface SearchResultProps {
    category?: string
    title: string
    summary: string
    url: string
    data?: any
    iconUrl?: string
}

export class SearchResultCard extends React.Component<SearchResultProps, any> {

    resolveType() {
        const { data } = this.props;

        let typeName;
        if (data) {
            if (data.getIn(['dataClass', 'name'])) {
                typeName = data.getIn(['dataClass', 'name']);
            }
            else if (data.getIn(['sampleSet', 'name'])) {
                typeName = data.getIn(['sampleSet', 'name']);
            }

            if (typeName) {
                return (
                    <div>
                        <strong>Type: </strong>
                        {typeName}
                    </div>
                )
            }
        }
    }

    resolveImage() {
        const { category, data, iconUrl } = this.props;

        if (iconUrl) {
            return <img className="search-result__card-icon" src={iconUrl}/>
        }

        let iconSrc = '';
        if (data) {
            if (data.getIn(['dataClass', 'name'])) {
                iconSrc = data.getIn(['dataClass', 'name']).toLowerCase();
            }
            else if (data.getIn(['sampleSet', 'name'])) {
                const sampleSetName = data.getIn(['sampleSet', 'name']).toLowerCase();

                switch (sampleSetName) {
                    case SCHEMAS.SAMPLE_SETS.RAW_MATERIALS.queryName.toLowerCase():
                        iconSrc = 'ingredients';
                        break;
                    case SCHEMAS.SAMPLE_SETS.MIXTURE_BATCHES.queryName.toLowerCase():
                        iconSrc = 'batch';
                        break;
                    default:
                        iconSrc = 'samples';
                        break;
                }
            }
            else if (data.get('type') === 'sampleSet') {
                iconSrc = 'sample_set';
            }
            else if (data.get('type')) {
                iconSrc=data.get('type').toLowerCase();
            }
        }
        if (!iconSrc && category) {
            switch (category)
            {
                case 'workflowJob':
                    iconSrc = 'workflow';
                    break;
                case 'material':
                    iconSrc = 'samples';
                    break;
            }
        }

        return <SVGIcon
            iconDir={'_images'}
            iconSrc={iconSrc}
            className="search-result__card-icon"/>
    }

    render () {
        const { title, summary, url } = this.props;

        return (
            <a href={url}>
                <div className="row search-result__card-container">
                    <div className="col-md-2 hidden-sm hidden-xs search-result__card-icon__container">
                        {this.resolveImage()}
                    </div>
                    <div className="col-md-10 col-sm-12">
                        <div>
                            <h4 className="text-capitalize">
                                {title}
                            </h4>
                        </div>
                        {this.resolveType()}
                        <div title={summary}>
                            <strong>Summary: </strong> {summary.length ? ( summary.length <= 35 ? summary : summary.substr(0, 35) + "..." ): 'No summary provided'}
                        </div>
                    </div>
                </div>
            </a>
        )
    }
}
