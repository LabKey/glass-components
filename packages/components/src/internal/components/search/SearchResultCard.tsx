/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { FC, memo } from 'react';

import { SVGIcon } from '../base/SVGIcon';

import { SearchResultCardData } from './models';

interface SearchResultProps {
    cardData: SearchResultCardData;
    iconUrl?: string;
    summary: string;
    url: string;
}

interface DetailProps {
    label: string;
    title?: string;
    value: string;
}

const CardDetail: FC<DetailProps> = memo(({ label, title, value }) => (
    <div className="search-result__card-detail" title={title}>
        <strong>{label}: </strong>
        {value}
    </div>
));

export const SearchResultCard: FC<SearchResultProps> = memo(({ cardData, iconUrl, summary, url }) => {
    const { altText, category, iconDir, iconSrc, title, typeName } = cardData;
    let summaryText = 'No summary provided';

    if (summary.length) {
        summaryText = summary.length <= 35 ? summary : summary.substr(0, 35);
    }

    return (
        <a href={url}>
            <div className="row search-result__card-container">
                <div className="hidden-xs search-result__card-icon__container">
                    {iconUrl && <img className="search-result__card-icon" src={iconUrl} />}
                    {!iconUrl && (
                        <SVGIcon
                            iconDir={iconDir}
                            iconSrc={iconSrc}
                            alt={altText}
                            className="search-result__card-icon"
                        />
                    )}
                </div>
                <div className="col-sm-12">
                    <div className="col-sm-12 search-result__title">
                        <h4 className="text-capitalize">{title}</h4>
                        {category && <div className="search-result__category-badge">{category}</div>}
                        {typeName && <div className="search-result__type-badge">{typeName}</div>}
                    </div>
                    <div className="search-result__summary">{summaryText}</div>
                </div>
            </div>
        </a>
    );
});
