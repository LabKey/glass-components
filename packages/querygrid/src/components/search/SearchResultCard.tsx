/*
 * Copyright (c) 2016-2018 LabKey Corporation. All rights reserved. No portion of this work may be reproduced in
 * any form or by any electronic or mechanical means without written permission from LabKey Corporation.
 */
import * as React from 'react'
import {SVGIcon, SCHEMAS} from "@glass/base";

interface SearchResultProps {
    title: string
    summary: string
    url: string
    data?: any
}

export class SearchResultCard extends React.Component<SearchResultProps, any> {

    resolveType() {
        const { data } = this.props;

        let typeName;
        if (data) {
            if (data.getIn(['dataClass', 'name'])) {
                typeName = data.getIn(['dataClass', 'name']);
            }
            else if (data.getIn(['sampleSet', 'name'])) {
                typeName = data.getIn(['sampleSet', 'name']);
            }

            if (typeName) {
                return (
                    <div>
                        <strong>Type: </strong>
                        {typeName}
                    </div>
                )
            }
        }
    }

    resolveImage() {
        const { data } = this.props;

        let iconSrc = '';

        if (data) {
            if (data.getIn(['dataClass', 'name'])) {
                iconSrc = data.getIn(['dataClass', 'name']).toLowerCase();
            }
            else if (data.getIn(['sampleSet', 'name'])) {
                const sampleSetName = data.getIn(['sampleSet', 'name']).toLowerCase();

                switch (sampleSetName) {
                    case SCHEMAS.SAMPLE_SETS.RAW_MATERIALS.queryName.toLowerCase():
                        iconSrc = 'ingredients';
                        break;
                    case SCHEMAS.SAMPLE_SETS.MIXTURE_BATCHES.queryName.toLowerCase():
                        iconSrc = 'batch';
                        break;
                    default:
                        iconSrc = 'samples';
                        break;
                }
            }
            else if (data.get('type')) {
                iconSrc=data.get('type').toLowerCase();
            }
        }

        return <SVGIcon
            iconDir={'_images'}
            iconSrc={iconSrc}
            className="search-result__card-icon"/>
    }

    render () {
        const { title, summary, url } = this.props;

        return (
            <a href={url}>
                <div className="row search-result__card-container">
                    <div className="col-md-2 hidden-sm hidden-xs search-result__card-icon__container">
                        {this.resolveImage()}
                    </div>
                    <div className="col-md-10 col-sm-12">
                        <div>
                            <h4 className="text-capitalize">
                                {title}
                            </h4>
                        </div>
                        {this.resolveType()}
                        <div title={summary}>
                            <strong>Summary: </strong> {summary.length ? ( summary.length <= 35 ? summary : summary.substr(0, 35) + "..." ): 'No summary provided'}
                        </div>
                    </div>
                </div>
            </a>
        )
    }
}
