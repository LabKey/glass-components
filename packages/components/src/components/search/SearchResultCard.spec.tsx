/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react'
import renderer from 'react-test-renderer'
import { fromJS } from 'immutable'
import { SearchResultCard } from "./SearchResultCard";
import { mount } from "enzyme";

describe("<SearchResultCard/>", () => {

    test("default props", () => {
        const component = (
            <SearchResultCard title={'Card Title'} summary={'Card Summary'} url={'#card'}/>
        );

        const wrapper = mount(component);
        const icon = wrapper.find('img');
        expect(icon.getDOMNode().getAttribute('src')).toBe('/labkey/_images/default.svg');
        expect(wrapper.text().indexOf('Type: ')).toBe(-1);
        wrapper.unmount();

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("resolve image and type from dataClass", () => {
        const data = fromJS({dataClass: {name: 'molecule'}});
        const component = (
            <SearchResultCard title={'Card Title'} summary={'Card Summary'} url={'#card'} data={data}/>
        );

        const wrapper = mount(component);
        const icon = wrapper.find('img');
        expect(icon.getDOMNode().getAttribute('src')).toBe('/labkey/_images/molecule.svg');
        expect(wrapper.text().indexOf('Type: molecule')).toBeGreaterThan(-1);
        wrapper.unmount();
    });

    test("resolve image and type from sampleSet", () => {
        const data = fromJS({sampleSet: {name: 'Sample Set 1'}});
        const component = (
            <SearchResultCard title={'Card Title'} summary={'Card Summary'} url={'#card'} data={data}/>
        );

        const wrapper = mount(component);
        const icon = wrapper.find('img');
        expect(icon.getDOMNode().getAttribute('src')).toBe('/labkey/_images/samples.svg');
        expect(wrapper.text().indexOf('Type: Sample Set 1')).toBeGreaterThan(-1);
        wrapper.unmount();
    });

    test("with iconUrl", () => {
        const iconUrl = 'http://labkey.wpengine.com/wp-content/uploads/2015/12/cropped-LK-icon.png';
        const component = (
            <SearchResultCard title={'Card Title'} summary={'Card Summary'} url={'#card'} iconUrl={iconUrl}/>
        );

        const wrapper = mount(component);
        const icon = wrapper.find('img');
        expect(icon.getDOMNode().getAttribute('src')).toBe(iconUrl);
        expect(wrapper.text().indexOf('Type: ')).toBe(-1);
        wrapper.unmount();
    });

});
