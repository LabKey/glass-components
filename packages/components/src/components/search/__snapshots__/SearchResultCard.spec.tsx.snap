// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultCard/> default props 1`] = `
<a
  href="#card"
>
  <div
    className="row search-result__card-container"
  >
    <div
      className="col-md-2 hidden-sm hidden-xs search-result__card-icon__container"
    >
      <img
        alt="-icon"
        className="search-result__card-icon"
        height="100%"
        src="/labkey/_images/default.svg"
        width="100%"
      />
    </div>
    <div
      className="col-md-10 col-sm-12"
    >
      <div>
        <h4
          className="text-capitalize"
        >
          Card Title
        </h4>
      </div>
      <div
        title="Card Summary"
      >
        <strong>
          Summary: 
        </strong>
         
        Card Summary
      </div>
    </div>
  </div>
</a>
`;
